### ISSUE TITLE
state of v1 submissions after they got transferred to v2?

### ISSUE THREAD

**Issue by NobbZ** [exercism/v2-feedback#190]
What will be the state of submissions we already did on v1 after they got transfered to v2 in the upcoming and final release?

I do currently see these options:

* Approved and published: This way they have basically the same state as before. Everyone could in theory see and comment on them. But the user isn't able to get mentors feedback. Also those solutions might be below par, such that other students might pick up inidiomatic or even dangerous habits from them.
* Approved and unpublished: This way a user won't get mentors feedback but were allowed to publish his solution at will to other people who could "learn" from them as in the former bullet point.
* In mentors queue: This is what we did August 2017 and produced a year spanning menoring backlog. This would put a lot of burden on the mentors to mentor solutions from people that probably do not even use exercism anymore or never wanted feedback. After mentors got through the backlog to current exercises they might be exhausted and quality of feedback might have decreased massively.
* In independent mode: This way no one could see those exercises, neither would a mentor see them. If we allow students to selectively move solutions from independant mode to mentor mode (at a limited rate), students that want to were able to get valuable feedback. If we do not allow that, then it is as if those submissions never had happened, while they also forbid proper participation in that track forever.

My favorite were the latter with "allow to move to mentor mode".

Based on slack: https://exercism-mentors.slack.com/archives/CAQP7JL3T/p1530199566000487

**Comment by iHiD** [comment]
Thanks @NobbZ. So my gut reaction is that they should be "Approved and published". @kytrinyx Could you confirm what they are atm pls? We can discuss this with the product team too.

**Comment by NobbZ** [comment]
> So my gut reaction is that they should be "Approved and published".

Can you explain why? I consider this by far the worst, as this will break everything mentorship is considered to give us.

**Comment by rpottsoh** [comment]
How about not transferring any solutions from V1 to V2?  Those students that are still active may then re-submit their solutions.  I guess there is potential loss of value from those solutions that do not get re-submitted due to the absence of the student that had submitted it originally to V1.

**Comment by iHiD** [comment]
> Can you explain why? I consider this by far the worst, as this will break everything mentorship is considered to give us.

> This way they have basically the same state as before.

Because of this. I think that the work people have already done should remain in a state that's expected for them. Things were effectively approved and published before. If they remain unapproved then someone won't be able to progress without waiting for a mentor, and we end up with a 4yr backlog of solutions to work through. If they remain unpublished, then we break people's previous links.

> If we allow students to selectively move solutions from independant mode to mentor mode (at a limited rate), students that want to were able to get valuable feedback.

This is a nice idea, but the tech isn't set up like that. A whole track is either in individual mode or mentored mode, and it would be too complicated to change that now.

> And if all my former submissions are already approved, how can I then selectively bring them to a moderators attention? Especially this is something I am waiting for. _(From Slack)_

This is a good point, and I'd like to work out how to do this. It would be nice if we could let you unapprove or flag a solution to go to the queue. I'll have a think and look at the codebase and work out if that is feasible.

> On day 1 of the new version of the site, will everyone be in "independent mode" by default? If not, how will that affect progress through exercises? _(from slack)

We should give existing users the option on a track to choose independent mode or mentored mode for future submissions to that track.

**Comment by iHiD** [comment]
I think we agree on what we want to achieve: 
- No old burden for mentors
- No major unexpected changes for users
- Not being blocked by everything needing approval 
- The ability to get mentoring on old solutions

I think the question is just how we achieve that best in the UI.

**Comment by iHiD** [comment]
>  I guess there is potential loss of value from those solutions that do not get re-submitted due to the absence of the student that had submitted it originally to V1.

Yeah - exactly. I think it'll be a real loss to lose those.

**Comment by NobbZ** [comment]
You forget about one thing: Not have unidiomatic or dangerous code publicly available.

Also whats about the history? As I understand it, the current v2 does only publish the latest approved submission, not the previous one.

**Comment by iHiD** [comment]
> You forget about one thing: Not have unidiomatic or dangerous code publicly available.

That's already part of v1 though? I don't feel like we're changing anything here?

> Also whats about the history? As I understand it, the current v2 does only publish the latest approved submission, not the previous one.

The history is being carried over, and you can still look at previous iterations of your own solutions, but we decided not to show previous iterations publicly, as we think that is best private between the learner/mentor. That's not a decision we were certain on, but we felt like it was a good first default to have, and something to reevaluate later.

**Comment by NobbZ** [comment]
> That's already part of v1 though? I don't feel like we're changing anything here?

But hasn't menorship beeing praised as a solution to this problem in V1? Pieces of dangerous code that are uncommented in the public and could be taken by new users of a language as the way to do?

> The history is being carried over, and you can still look at previous iterations of your own solutions, but we decided not to show previous iterations publicly, as we think that is best private between the learner/mentor

But often (especially for v1 submissions) the old submissions have valuable discussions attached to them. Thats not a mentor thing, that discussions have taken place publicly.

Also the last iteration (at least for me) is not always the one I'd like to publish if I had a choice.

An example for myself is this ocaml piece, http://exercism.io/submissions/dfc845ad38c14f4683721a7f39d3e7d7, if I had to choose to publish only one submission, I'd publish iteration 3.

But to be honest, comments on iteration 4 are valuable, as they might make me re-iterate to try that `Seq` module at a certain point in time.

Also just as a side note, your comments so far do suggest that re-iterations after approval won't be possible anymore. This is one of the major fun parts for me. Comming back after years and re-iterate old exercises. Just take a look at my recent activity.

**Comment by kytrinyx** [comment]
Hi @NobbZ, these are all fantastic questions.

The short answer is:
- we made some decisions
- we have deferred other decisions until after the launch
- we may change our minds about things we already decided as soon as we see the new site in action

**Timescale**
For the first pass, anything newer than two weeks remains in the mentor queue. There are a *lot* of old submissions that people don't come back to. Anything older than two weeks will be marked as approved by the migration bot, and the final iteration will be published.

**Reopening published solutions**
I am quite certain that we want to make it possible to go back to and re-open old exercises. It's a fantastic way to learn and apply new lessons. We will need to do a bit of design around how and when this is done. Additionally, we have discussed at length how to make it so that you can do an exercise _over again_, or try different approaches to the same exercise in parallel. These will all require product discussions and design before we can tackle them.

**Quality of published solutions**

Approving and publishing a solution doesn't mean that it's exemplary, it means that someone has completed an exercise (for some value of completed) and learned something from it. It's an interesting datapoint, but it is not meant to be representative of any particular standard of excellence or bar of quality. The key deliverable is _learning_.

I think it's crucial to remember that while Exercism as a whole is about becoming fluent in (among other things) the idioms of a language, it is not fundamentally _about_ good engineering practices. Good programming habits and patterns are a side effect. The core of Exercism is about supporting people's learning along the way. There are typically over 100 exercises, which means that for any particular idea, there will likely be many opportunities to discuss it, and often the final solution of an exercise is not the ultimate best-possible solution. That's why it's so great to go back after a year or two and give it another go!

**Value of the corpus of 5 years of data**

People learn a lot from browsing exercises, even when the code is not exemplary. We did not want to take that away from people.

It would probably be worth making the distinction somehow in the UI that these exercises came through a different process than the current mentorship process. It would be fairly straight-forward to do, I think. I'll bring it to the product team to discuss.

**Which iteration to publish**

We had to make a decision programmatically, so we chose the last iteration. I completely understand the desire to publish a different iteration, and I think that especially for people who are curating their profile, it would be crucial to be able to publish whatever communicates the most interesting or important thing for them. This is something we'll bring to the product team to discuss.

**Historical discussions**

Here there is a very strong tension between (a) it wasn't technically mentoring because we didn't have mentoring, and (b) the discussions were a part of the learning process, and can therefore be classified as mentoring. No matter which we choose, we lose something.

We have chosen to classify the historical discussions on all but the last iteration as a mentorship discussion because it's as close as we can get to the model that we have without making things extraordinarily complicated. I completely agree that we lose _something_, sometimes, when there were particularly interesting discussions.

Everyone who participated in the discussion will have access to it on the new site.

I think there is some possibility here of considering ways of bringing these interesting discussions into the light, but I don't think that this will be in the form of product features on the new site.

The most interesting discussions seem to be in two categories (if you can think of others, I'd love to hear about them!)

1. fantastic mentoring -- supportive guidance with intriguing insights along the way
1. peer-to-peer explorations -- a delightful deep dive into the intricacies of a concept

The first type of conversation is something that we will be bringing into the mentor resources and guides. Not word for word, but we intend to find these conversations and make the essence of them available.

The second type of conversation is not the core purpose of Exercism. It's a delightful serendipitous byproduct of having thoughtful, caring, enthusiastic people get together and talk about some of their favorite subjects. I think that the best use for these would be for the people who participated to use these as fodder for blog posts, podcasts, books, conference talks, and the like.

# 
While none of this is set in stone, we have put a lot of thought into the trade-offs, and tried to be true to both the history of Exercism, the expectations of those who have used it, and the new model. In some places we've accepted that there is no perfect solution, and that we're OK with something that will be _good enough for the majority of the cases_.

**Comment by ErikSchierboom** [comment]
> I am quite certain that we want to make it possible to go back to and re-open old exercises.

I would greatly be in favor of this, as this is precisely what I periodically do.

> It would probably be worth making the distinction somehow in the UI that these exercises came through a different process than the current mentorship process.

Excellent idea!

**Comment by guygastineau** [comment]
@kytrinyx thank you for the update on the implementation of V2.

Also, in response to the threat of unidiomatic or 'dangerous' code being published and taken as a good practice by learners.

Personally I found reading solutions and analyzing them to discover whether or not they we're implemented in idiomatic ways or with best practices in mind helped me learn.  Also, hopefully no one working on a mission critical part of a project is trolling solutions to build their codebase.

I intend this post as a constructive perspective and do not intend a negative attitude.

**Comment by kytrinyx** [comment]
_This repo is being deprecated. We've imported this issue to the https://github.com/exercism/exercism.io repository._
